src/conn_state.c

```diff
@@ -0,0 +1,163 @@
+#include <fcntl.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <arpa/inet.h>
+#include <net/if.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+
+#include "context.h"
+#include "log.h"
+
+static
+int
+connect_udp_socket(char* udp_address, int udp_length, struct addrinfo* connect_addr)
+{
+    int rc = 0;
+    struct sockaddr_in udp_sock_addr;
+    int udp_sockfd = -1;
+
+    switch(connect_addr->ai_family){
+        case AF_INET:
+            udp_sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+            break;
+        case AF_INET6:
+            udp_sockfd = socket(AF_INET6, SOCK_DGRAM, 0);
+            break;
+    }
+
+    if (udp_sockfd < 0) {
+        enftun_log_error("UDP: Failed to create socket: %s\n", strerror(errno));
+    }
+
+    socklen_t udp_addr_len = sizeof(struct sockaddr_in);
+
+    memset(&udp_sock_addr, 0, udp_addr_len);
+    udp_sock_addr.sin_addr.s_addr = htonl(INADDR_ANY);
+    udp_sock_addr.sin_port = htons(0);
+    udp_sock_addr.sin_family = connect_addr->ai_family;
+
+    rc = connect(udp_sockfd, connect_addr->ai_addr, connect_addr->ai_addrlen);
+    if (0 != rc){
+        close(udp_sockfd);
+        return -1;
+    }
+
+    rc = getsockname(udp_sockfd, (struct sockaddr*)&udp_sock_addr, &udp_addr_len);
+    if (0 != rc){
+        close(udp_sockfd);
+        return -1;
+    }
+
+    inet_ntop(connect_addr->ai_family, (struct sockaddr_in*)&udp_sock_addr.sin_addr, udp_address, udp_length);
+
+    close(udp_sockfd);
+
+    return 0;
+}
+
+static
+int
+udp_check(struct enftun_conn_state* conn_state)
+{
+    struct enftun_context* ctx = (struct enftun_context*)conn_state->data;
+    char udp_address[45];
+    connect_udp_socket(udp_address, sizeof(udp_address), &ctx->tls.sock.local_addr);
+
+    struct sockaddr_in local_addr;
+    socklen_t addr_len = sizeof(struct sockaddr_in);
+
+    memset(&local_addr, 0, addr_len);
+    local_addr.sin_addr.s_addr = htonl(INADDR_ANY);
+    local_addr.sin_port = htons(0);
+    local_addr.sin_family = ctx->tls.sock.local_addr.ai_family;
+
+    char local_ip[45];
+    getsockname(ctx->tls.sock.fd, (struct sockaddr*)&local_addr, &addr_len);
+    inet_ntop(ctx->tls.sock.local_addr.ai_family, (struct sockaddr_in*)&local_addr.sin_addr, local_ip, sizeof(local_ip));
+
+    if (0 == strcmp(udp_address, local_ip)) {
+        return 0;
+    }
+    else {
+        enftun_log_info("UDP check found a higher priority route. Reconnecting...\n");
+        return -1;
+    }
+
+    return 0;
+}
+
+static
+void
+check_conn_state(struct enftun_conn_state* conn_state)
+{
+    int rc = udp_check(conn_state);
+    if (0 != rc) {
+        conn_state->reconnect(conn_state);
+    }
+}
+
+static
+void
+on_poll(uv_poll_t* handle, int status, int events)
+{
+    (void) events;
+    struct enftun_conn_state* conn_state = handle->data;
+
+    if (status < 0)
+        return;
+    if (0 == status)
+        check_conn_state(conn_state);
+}
+
+int
+enftun_conn_state_start(struct enftun_conn_state* conn_state,
```

Comment:
I led you astray yesterday, suggesting to put the `netlink_connect` and the `uv_poll_start` in the `conn_state_start` function.  Because now we have one start function, but two opposites  (a `stop` and a `close`).

One principle in this code base is that each `_init_` is matched by a `_free`, each `_connect` is matched by a `_close`, and each `_start` is matched by a `_stop`.  Ideally, both in the pair are called in the same function, making it easy to visually inspect the code to see that everything is being cleaned up properly.

The `close` and `stop` need to be separated, so I suggest separating the `start` into `_prepare` (connect the netlink socket) and `_start` (start the uv poll).